src/mscorlib/src/System/Diagnostics/Stacktrace.cs

```diff
@@ -664,6 +681,37 @@ private static bool ShowInStackTrace(MethodBase mb)
             return !(mb.IsDefined(typeof(StackTraceHiddenAttribute)) || (mb.DeclaringType?.IsDefined(typeof(StackTraceHiddenAttribute)) ?? false));
         }
 
+        private static bool TryResolveStateMachineMethod(ref MethodBase mb, ref Type declaringType)
+        {
+            Debug.Assert(mb != null);
+            Debug.Assert(declaringType != null);
+
+            Type dt = declaringType.DeclaringType;
```

Comment:
>Type dt = declaringType.DeclaringType [](start = 12, length = 37)

This makes an assumption that state machine type is a nested type to the type that defines the kickoff method. That's not something the compiler guarantees, but I think it's not unreasonable to use that fact as a heuristic, especially since the code is already well hardened against not being able to find the method.

@jaredpar @vsadov To confirm that this is ok.

Reply:
I was thinking as a follow up; it could check pdb first; then use this approach as a fallback - though the relationship with `System.Diagnostics.StackTraceSymbols.GetSourceLineInfo` is a little indirect

Reply:
I would avoid the PDB for this purpose. We've got some past experience from .Net Native that depending on a PDB + another mechanism, both of which have independent failure conditions, is sufficiently complex and unexpected that users are confused. If we don't have enough info in the binary to do the resolution correctly I'd encourage changing the C# compiler so that it provides more.

Reply:
The pointers the other way would need to be to a Method(Info/Base) which makes it more tricky. The statemachine is a Type so its easier to mark. Same issue for lambdas and local functions - which would be method <-> method pointers.

This change with the previous one https://github.com/dotnet/coreclr/pull/14652 address the majority of complaints I've seen about stack traces - other than a desire for pdbs to be move widely available; but I think that's also being addressed.

Reply:
> I would avoid the PDB for this purpose. We've got some past experience from .Net Native that depending on a PDB + another mechanism, both of which have independent failure conditions, is sufficiently complex and unexpected that users are confused

I'm curious what exactly were users confused about in the .NET Native case. I don't think anyone is confused about the fact that stack trace has file names and line numbers when PDB is present and doesn't when it's not. Formatting stack trace in a way that's closer to source when the source location (file name and line number) is included makes sense to me. 

We can certainly add more information to metadata. It's a trade off though. More metadata means the assembly is bigger and the extra data is only gonna be used for diagnostics and debugging. That's what we have standalone PDBs for. Since the state machine information is already contained in the PDB we would have two different representations of the same data stored twice.

Reply:
Where would be a good place to raise this type of more general issue?

More information and how to use that information? As its a cross cutting concern that would need some design and could effect many parts (runtime, compilers, apis) e.g. rosyln/csharplang/corefx/core etc

Reply:
We are diving a bit down the rabbit hole... I think the easiest solution is that the Roslyn team decide that they will always use a nested type and we can leave this as it is. I agree its not currently a guarantee the compiler makes, but naively it doesn't seem like a huge burden to commit to that. If it ever changed a bunch of IL disassemblers would probably break, the Roslyn team would have to spend time changing and retesting a feature that already works, and its not clear any value would be gained.



However to answer the questions asked, In .Net Native we first checked if a method had metadata, and then we did a fallback to PDB, and if neither of those worked then we rendered the stack trace frame as an RVA. Users would see stack traces that looked like this:
Foo.Bar()
MyApp.exe+0x1235123
SomeOtherTypeInMyApp.Baz()
Type3.DoSomething()

Users are pretty familiar with the idea that PDBs have source mapping in it so if that is present/missing they tend to correctly guess its because of a PDB. But when it comes to figuring out why one of those frames was an RVA and the others were properly named they have no clue. The answer is that due to esoteric details of the .Net Native compiler some methods had metadata and others did not.

You can map that onto the current proposal by switching 'esoteric details of the .Net Native compiler' to 'esoteric details of the C# compiler', and switch 'method rendered with RVA' to 'method rendered with untransformed async state machine'. The degree to which they dislike an RVA might be different, but the fact that the behavior will seem unexplainable and confusing I'd expect to be consistent.

> We can certainly add more information to metadata. It's a trade off though. More metadata means the assembly is bigger and the extra data is only gonna be used for diagnostics and debugging

In .Net Native customer feedback around improving stack traces was one of the most common complaints of the entire platform. Customers there would gladly accept non-trivial increases in binary size if it meant higher quality stack traces. I agree its a case by case basis, but if you tell me you need +2% IL binary size to add the mapping I'd vote go for it (after first asking why do we care so much about preserving this particular degree of freedom for the compiler that we are willing to pay any size cost at all)

Reply:
> We are diving a bit down the rabbit hole... 

I mean raising an issue for other items rather than particularly this PR. The other items that come out quite weird in stack traces are lambdas and local functions and there's not currently a convenient way of tying them back to the source method; or an obvious way to name them. 

Though local functions could use the `+Method` like for the iterator; but lambdas have no name.

Reply:
I also don't think they trouble people as much as async and iterators; but local functions seem to be quite a loved feature - so the flip side of confusing stack traces may follow

Reply:
> If it ever changed a bunch of IL disassemblers would probably break

Every new version of the compiler breaks IL disassemblers in some way. The exact shape of metadata marked by CompilerGenerated attribute or specific IL patterns are not something we guarantee to not change.

> its not clear any value would be gained.

We change these things whenever we identify some optimization we can do. Or a new language feature implementation requires other existing feature that it interacts with to be changed.

Reply:
@benaadams Roslyn repo would be the right one to discuss such issues.

Reply:
@noahfalk I totally understand the customers who complain about RVAs. That's a completely opposite direction than we are doing here though. Users were used to see type and method names in the stack traces. .NET Native sometimes replaced them with a number that has no meaning to anyone. What we are proposing here is the opposite direction. Take compiler implementation detail that users don't see in their source code, but historically are required to understand in order to interpret stack traces, and present it in a way that matches the source code. Yes, it's a change and some may be wondering what's going on. However, since the stack trace would display the names exactly as they are in the source code I don't see anyone having an issue figuring out what do they mean. In fact, a newcomer who doesn't have a previous understanding of the said compiler implementation details will have absolutely no problem understanding it. They will only be wondering what's going on when the PDB is not present. But that's not making things worse than they are today (if we can't figure it out from metadata).

I think the best approach that works without changing the compiler is: 
1) if PDB is present use the information there -- this information will always be precise. We also won't need to enumerate all methods checking for the attribute, making it more efficient.
2) if PDB is not present fall back to the code that we have in this PR today. That will work fine until we need to change something in the compiler, which is unlikely.

Reply:
@tmat - Not a fan of that plan I'm afraid. This is how I imagine it playing out in my head in a few years:

Dev: Yey these stack traces are nice
Microsoft: Here you go, new version of VS is even better! 
Dev: Cool! ... wait why are my async stack traces sometimes worse, what is going on?
[ bunch of web searching, some people say its the compiler version that matters, others say its PDB, its confusing why different causes are blamed ]
Microsoft: We changed the compiler sorry, but don't worry if you put your PDB next the binary then you can recover the behavior you had before.
Dev: Seriously? I need to pay a +40% size on disk cost and rework all my deployment logic to keep PDBs next to binaries now? Clearly the compiler was able to make this before without a PDB, why is it my problem now? Microsoft I don't care if the binaries are 2% smaller now or that you added a new C# feature I don't use in my code, just give me my old binary that worked how I wanted. Does my company policy even allow PDBs to be on my production machines? Oh I found an easier solution... reverting back to my old version of the .Net SDK and staying there until this is fixed. 


If the compiler doesn't want to guarantee that nested classes will be used in the future than I would counter-propose this plan:

1. Define a new attribute/resource/anything that has the data allowing the stacktrace code to map the statemachine method name back to the source method name. 
2. The runtime will search for that attribute/resource/whatever first and then if it isn't found it will do what the code does now.
3. This allows the compiler to avoid generating any extra data as long as it sticks to the current plan. If the compiler ever wants to change course it can start generating the extra data to preserve the existing runtime behavior.

Reply:
@noahfalk Good idea.

Reply:
@noahfalk So I think the plan now is to check with @jaredpar if we could guarantee that SM is nested next to the kickoff method. If we don't want to do that then we define this attribute for future use.

Reply:
Guarantee would be that the attribute (and runtime to check for it); or some other mechanism, would be implemented prior to making a change to the SM nesting?

Or should it be implemented first so its "baked in" early?

Reply:
If the compiler team is unwilling to promise the current behavior I'd bake it early, otherwise leave it as-is and the compiler promises to maintain compat.

Reply:
Raised a discussion issue https://github.com/dotnet/roslyn/issues/22857

Reply:
Is also issue for aspnet who process their own stack frames (and translate generics to C#); related PR  https://github.com/aspnet/Common/pull/274